### PR TITLE
UW-407

### PR DESCRIPTION
**Pull request by maddenp-noaa**
**Description**

Factor string literals describing actions to take when operations on existing directories are attempted out to a dataclass shared across modules. Update use sites and related tests. DRY out `handle_existing()`.

**Type of change**

- [x] Maintenance

**How Has This Been Tested?**

`make test` & Actions

**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

### CODE REVIEW

File: src/uwtools/utils/file.py [hunk 1]
```diff
@@ -166,6 +156,21 @@ def resource_pathobj(suffix: str = "") -> Path:
...
+
+    :param exist_act: Action to check.
+    :param valid_actions: Actions valid for the caller's context.
+    :raises: ValueError if specified action is invalid.
+    """
+    if exist_act not in valid_actions:
+        raise ValueError(
+            'Specify one of %s as exist_act, not "%s"'
+            % (", ".join(f'"{x}"' for x in valid_actions), exist_act)
+        )
```
Comment (maddenp-noaa) [inline src/uwtools/utils/file.py:171]: Similar action-validation logic is needed in two modules, so factor it out here.

File: src/uwtools/drivers/forecast.py [hunk 2]
```diff
@@ -52,25 +52,26 @@ def batch_script(self) -> BatchScript:
...
+    ) -> None:
         """
         Collects the name of the desired run directory, and has an optional flag for what to do if
         the run directory specified already exists. Creates the run directory and adds
         subdirectories INPUT and RESTART. Verifies creation of all directories.
 
         :param run_directory: Path of desired run directory.
-        :param exist_act: Could be any of 'delete', 'rename', 'quit'. Sets how the program responds
-            to a preexisting run directory. The default is to delete the old run directory.
+        :param exist_act: Action when run directory exists: "delete" (default), "quit", or "rename"
```
[the remark this answers is not in the document]
Comment (WeirAE) [reply src/uwtools/drivers/forecast.py:63]: I've been pondering this one a bit, but I'm kind of stuck at your conclusion. My preference leans towards allowing string literals and leaving the docstring, but I could be swayed

File: src/uwtools/drivers/driver.py [hunk 3]
```diff
@@ -81,13 +81,14 @@ def run_cmd(self) -> str:
...
-        args_str = " ".join(str(arg) for arg in run_time_args)
-        return f"{run_cmd} {args_str} {exec_name}"
+        :return: Collated string that contains MPI command, runtime arguments, and exec name.
+        """
+        components = [
+            self._platform_config.get("mpicmd"),  # MPI run program
+            *[str(x) for x in self._config["runtime_info"].get("mpi_args", [])],  # MPI arguments
+            self._config["exec_name"],  # NWP tool executable name
+        ]
+        return " ".join(filter(None, components))
```
Comment (maddenp-noaa) [inline src/uwtools/drivers/driver.py:91]: See [previous discussion](https://github.com/ufs-community/workflow-tools/pull/327/files#r1380616545) for rationale for this change.

File: src/uwtools/drivers/forecast.py [hunk 4]
```diff
@@ -176,7 +172,8 @@ def run(self, cycle: datetime) -> bool:
         """
         Runs FV3 either locally or via a batch-script submission.
 
-        :return: Did the FV3 run exit with success status?
+        :param cycle: The forecast cycle to run.
```
Comment (maddenp-noaa) [inline src/uwtools/drivers/forecast.py:175]: Add missing parameter description.

File: src/uwtools/tests/drivers/test_forecast.py [hunk 5]
```diff
@@ -373,11 +374,14 @@ def test_FV3Forecast_run(fv3_run_assets, vals):
...
-            fcstobj._run_via_batch_submission()
-        execute.assert_called_once_with(cmd=ANY)
+    fcstobj = FV3Forecast(config_file=config_file, batch_script=batch_script)
+    with patch.object(fcstobj, "_config", config):
+        with patch.object(scheduler, "execute") as execute:
+            execute.return_value = ns(success=True)
+            success, lines = fcstobj._run_via_batch_submission()
+            assert success is True
+            assert lines[0] == "Batch script:"
+            execute.assert_called_once_with(cmd=ANY)
```
Comment (maddenp-noaa) [inline src/uwtools/tests/drivers/test_forecast.py:384]: I added a couple of assertions similar to the ones in the test just below, which is very similar.

File: src/uwtools/drivers/forecast.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (elcarpenterNOAA) [reply src/uwtools/drivers/forecast.py:63]: I'm not sure what the best solution for this is either. Consistency is important, but I would likely also vote for allowing string literals.

File: src/uwtools/utils/file.py [hunk 6]
```diff
@@ -91,35 +91,27 @@ def get_file_type(path: DefinitePath) -> str:
...
+        try:
+            if exist_act == ExistAct.delete:
+                shutil.rmtree(directory)
+            elif exist_act == ExistAct.rename:
+                save_dir = "%s%s" % (directory, dt.now().strftime("_%Y%m%d_%H%M%S"))
+                shutil.move(directory, save_dir)
+        except (FileExistsError, RuntimeError) as e:
+            msg = f"Could not {exist_act} directory {directory}"
+            log.critical(msg)
+            raise RuntimeError(msg) from e
```
[the remark this answers is not in the document]
Comment (elcarpenterNOAA) [reply src/uwtools/utils/file.py:112]: Much cleaner! I really like this change.

**Review by elcarpenterNOAA — APPROVED** [review]
LGTM. I appreciate the added precision in the docstrings and inline comments.

File: src/uwtools/tests/drivers/test_forecast.py (hunk 5 above)
Comment (NaureenBharwaniNOAA) [inline src/uwtools/tests/drivers/test_forecast.py:384]: This section cleaned up very nice

File: src/uwtools/types.py [hunk 7]
```diff
@@ -1,5 +1,17 @@
...
 from typing import Optional, Union
 
 DefinitePath = Union[Path, str]
 OptionalPath = Optional[DefinitePath]
+
+
+@dataclass
+class ExistAct:
+    """
+    Possible actions to taken when a directory already exists.
```
Comment (NaureenBharwaniNOAA) [inline src/uwtools/types.py:12]: Small typo I believe, taken --> take

Comment (maddenp-noaa) [reply src/uwtools/types.py]: Fixed, thank you.

**Review by NaureenBharwaniNOAA — APPROVED** [review]
Looks good!

**Pull request merged**